### PR TITLE
drwtsn32 improvements

### PR DESCRIPTION
**Pull request by wjk**
## Purpose

General quality-of-life improvements to drwtsn32.

JIRA issue: [CORE-14180](https://jira.reactos.org/browse/CORE-14180)

## Proposed changes

- drwtsn32 now saves to a location configurable in the Registry (or to the desktop if not specified)
- drwtsn32 attempts to write a minidump, if possible (this is disabled by default, and can be enabled using the Registry)
- drwtsn32 now has an RC file, including version information and a common-controls manifest

## TODO

- [x] Tested, and works great. Configurable output directory setting is respected, and a minidump is written if requested.

### CODE REVIEW

**Comment by learn-more** [comment]
Why not the desktop?

File: base/applications/drwtsn32/main.cpp [hunk 1]
```diff
@@ -233,14 +331,16 @@ int main(int argc, char* argv[])
...
     TerminateProcess(data.ProcessHandle, data.ExceptionInfo.ExceptionRecord.ExceptionCode);
 
+    std::wstring FilenameW(Filename);
     std::string Message = "The application '";
     Message += data.ProcessName;
     Message += "' has just crashed :(\n";
     Message += "Information about this crash is saved to:\n";
-    Message += Filename;
+    Message += std::string(FilenameW.begin(), FilenameW.end());
     Message += "\nThis file is stored on your desktop.";
```
Comment (ThFabba) [inline base/applications/drwtsn32/main.cpp]: This part is no longer true

File: base/applications/drwtsn32/main.cpp [hunk 2]
```diff
@@ -233,14 +331,16 @@ int main(int argc, char* argv[])
     }
 
     PrintBugreport(output, data);
+    if (Settings_GetShouldWriteDump() && HasPath) WriteMinidump(OutputPath.c_str(), data);
```
Comment (ThFabba) [inline base/applications/drwtsn32/main.cpp]: Two lines please

File: base/applications/drwtsn32/main.cpp [hunk 3]
```diff
@@ -194,19 +280,31 @@ int main(int argc, char* argv[])
...
     }
 
     GetLocalTime(&st);
 
-    if (SHGetFolderPathA(NULL, CSIDL_DESKTOP, NULL, SHGFP_TYPE_CURRENT, Buffer) == S_OK &&
-        SUCCEEDED(StringCchPrintfA(Filename, _countof(Filename), "Appcrash_%d-%02d-%02d_%02d-%02d-%02d.txt",
+    std::wstring OutputPath = Settings_GetOutputPath();
+    BOOL HasPath = (OutputPath.size() != 0);
+
+    if (!PathIsDirectoryW(OutputPath.c_str())) {
```
Comment (ThFabba) [inline base/applications/drwtsn32/main.cpp]: The rest of the file places the brace on its own line, please follow suit.

File: base/applications/drwtsn32/main.cpp [hunk 4]
```diff
@@ -143,49 +147,131 @@ int abort(FILE* output, int err)
     return err;
 }
 
-int main(int argc, char* argv[])
+std::wstring Settings_GetOutputPath(void)
 {
+    WCHAR Buffer[MAX_PATH + 1];
```
Comment (learn-more) [inline base/applications/drwtsn32/main.cpp]: It's easier to just do WCHAR Buffer[MAX_PATH + 1] = {0};

File: base/applications/drwtsn32/main.cpp [hunk 5]
```diff
@@ -143,49 +147,131 @@ int abort(FILE* output, int err)
...
+    ZeroMemory(Buffer, _countof(Buffer) * sizeof(WCHAR));
+    ULONG BufferSize = _countof(Buffer);
+
+    CRegKey key;
+    if (key.Open(HKEY_CURRENT_USER, L"SOFTWARE\\ReactOS\\Crash Reporter", KEY_READ) != ERROR_SUCCESS) {
+        goto use_default_path;
```
Comment (learn-more) [inline base/applications/drwtsn32/main.cpp]: I'd prefer this without goto.

File: base/applications/drwtsn32/main.cpp [hunk 6]
```diff
@@ -143,49 +147,131 @@ int abort(FILE* output, int err)
...
+    if (!DumpSucceeded) {
+        // According to MSDN, this value is already an HRESULT, so don't convert it again.
+        hr = GetLastError();
+    }
+
+    CloseHandle(hDumpFile);
+    return hr;
+}
+
+int __stdcall wWinMain(HINSTANCE, HINSTANCE, LPWSTR cmdLine, INT)
```
Comment (learn-more) [inline base/applications/drwtsn32/main.cpp]: why convert to WCHAR?
WINAPI instead of __stdcall

File: base/applications/drwtsn32/main.cpp (hunk 1 above)
Comment (learn-more) [reply base/applications/drwtsn32/main.cpp]: And if you are adding a resource file, you might as well make this a message resource.

File: base/applications/drwtsn32/main.cpp (hunk 4 above)
Comment (HBelusca) [reply base/applications/drwtsn32/main.cpp]: Actually the value number of MAX_PATH already counts for the terminating NULL.
And, it's possible to do: `WCHAR Buffer[whatever] = L"";`

File: base/applications/drwtsn32/main.cpp [hunk 7]
```diff
@@ -239,11 +263,12 @@ int __stdcall wWinMain(HINSTANCE, HINSTANCE, LPWSTR cmdLine, INT)
 
     TerminateProcess(data.ProcessHandle, data.ExceptionInfo.ExceptionRecord.ExceptionCode);
 
+    std::wstring FilenameW(Filename);
     std::string Message = "The application '";
     Message += data.ProcessName;
     Message += "' has just crashed :(\n";
     Message += "Information about this crash is saved to:\n";
-    Message += Filename;
+    Message += std::string(FilenameW.begin(), FilenameW.end());
```
Comment (HBelusca) [inline base/applications/drwtsn32/main.cpp]: I'm not sure how this is going to work: does the UNICODE -> ANSI conversion works OK?

File: base/applications/drwtsn32/main.cpp [hunk 8]
```diff
@@ -227,6 +228,15 @@ int __stdcall wWinMain(HINSTANCE, HINSTANCE, LPWSTR cmdLine, INT)
        }
     }
 
+    if (!PathIsDirectoryW(Buffer)) {
+        int res = SHCreateDirectoryExW(NULL, Buffer, NULL);
```
Comment (HBelusca) [inline base/applications/drwtsn32/main.cpp]: Or just call kernel32's CreateDirectory function?

File: base/applications/drwtsn32/main.cpp (hunk 7 above)
Comment (learn-more) [reply base/applications/drwtsn32/main.cpp]: This is one of the ways to convert wstring to string

File: base/applications/drwtsn32/main.cpp (hunk 8 above)
Comment (learn-more) [reply base/applications/drwtsn32/main.cpp]: That does not create directories recursive

File: base/applications/drwtsn32/main.cpp (hunk 6 above)
Comment (wjk) [reply base/applications/drwtsn32/main.cpp]: I converted to WCHAR because I needed to use `CommandLineToArgvW()`, which has no ASCII version.

**Comment by wjk** [comment]
@learn-more I changed drwtsn32 to put it in My Documents out of personal preference; I thought placing it there would look cleaner. If you prefer the desktop, I'll change it back.

**Comment by learn-more** [comment]
@wjk I chose the desktop because it is easily accessible, and I guessed users would find those files without any problem.

I am fine with a registry override, but I would prefer to keep the desktop as fallback / default location, unless someone else has a strong argument against that.

**Comment by HBelusca** [comment]
```
../../base/applications/drwtsn32/main.cpp: In function 'int wWinMain(HINSTANCE, HINSTANCE, LPWSTR, int)':
../../base/applications/drwtsn32/main.cpp:358:40: error: 'ATL::CString' has no member named 'c_str'
../../base/applications/drwtsn32/main.cpp:358:61: error: 'ATL::CString' has no member named 'c_str'
```

**Comment by wjk** [comment]
@learn-more I'll change it back, no problem.  
@HBelusca 🤦‍♂️ I got the two string types confused. I'll fix that soon. Thanks!

**Comment by wjk** [comment]
I'd like to apologize for the very sloppy programming that I have been submitting in this PR. I have had to work from a machine with no RosBE installed, and as such have been unable to build my changes to ensure that they compile.

**Comment by binarymaster** [comment]
Totally not a problem, good contributions always are very appreciated! :)

**Comment by learn-more** [comment]
@wjk fyi: Currently you are using 2 different emails to commit, one gmail and one masked by github.

**Comment by wjk** [comment]
@learn-more Argh! I've been using two different computers to commit, and I forgot the emails were different between them. Unfortunately, I don't know how to edit the Git email address once the commits have already been pushed.

Even worse, I corrupted my branch by incorrectly attempting to rebase it off of master (as evidenced by the commits from other users above). 😦 I don't know how to undo this mess. Could someone please help me?

**Comment by HBelusca** [comment]
The commit 528be19c7e974aaa0bae2db583f2ff2469e16ad1 is the latest one that is OK, and everything after it is your rebase problem.
So in your branch, do a:
```
git reset --hard 528be19c7e974aaa0bae2db583f2ff2469e16ad1
git push --force-with-lease
```
The first command requests git to revert everything back to the commit 528be19c7...
At this point you can check with `git log` that you have the correct history.
And the second command pushes your branch to your remote one, forcefully overwriting the history and any commits that have been done in between: this will "erase" the wrong commits.

**Comment by wjk** [comment]
@HBelusca Thank you, thank you, thank you! You seriously saved my bacon! 🎉 👏

**Comment by wjk** [comment]
@learn-more All commits should now use my gmail address. Thanks for pointing that out!

**Comment by learn-more** [comment]
Maybe the control panel applet should be in a separate PR?

**Comment by wjk** [comment]
@learn-more That's fine by me.

**Comment by learn-more** [comment]
I'm thinking we might want to squash a few commits before merging this.
Maybe group by logical features?

**Comment by HBelusca** [comment]
Hello, any news about the status of this PR?

**Comment by wjk** [comment]
@HBelusca Sorry for the lack of activity; I have been busy with other things. I have rebased the PR somewhat to get rid of all the "fix build" commits. If you or @learn-more have any other suggestions regarding rebasing, feel free to let me know. Otherwise, as far as I am concerned this PR is ready to merge. Thanks!

**Comment by learn-more** [comment]
Thanks!

**Pull request closed without merging**